### ISSUE TITLE
time series prediction with tpot ?

### ISSUE THREAD

**Issue by AnnamalaiNagappan** [rhiever/tpot#345]
## I have two columns, date colums with format %Y-%m-%d and a metric column which has values as type float. 
some thing like this

> date, metric
> 2016-01-01, 34
> 2016-01-02, 43
> 
> so on..

## I need to predict the metric column based on past date values.
Is there any example that shows the concept of time series with tpot.

Am sorry to raise this as a issue (since I could not find any examples), it's just a query.

**Comment by rhiever** [comment]
Yes, this should be feasible. If you convert the date into some numeric format, you can have a `TPOTRegressor` try to find a model that predicts `metric` as a function of `date_numeric`.

For example, you can convert `date` into `date_numeric` by counting the number of days since 0 AD:

```
date_numeric = [year] * 365 + [month] * 31 + [day]
```

(or maybe something less hacky than that -- there's Python libraries for doing similar things)

Then `date_numeric` is your feature and `metric` is your outcome. Create a `TPOTRegressor` and fit:

```
from tpot import TPOTRegressor

my_tpot = TPOTRegressor()
my_tpot.fit(features, outcome)
print(my_tpot.score(features))
```

etc.

You'd want to do some sort of [time series split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) before passing the data to TPOT, of course. Probably also want to use TimeSeriesSplit in TPOT's cv parameter, as discussed [here](https://github.com/rhiever/tpot/issues/342). Basically the same way you would approach a time series problem in scikit-learn.

**Comment by AnnamalaiNagappan** [comment]
Thank you for the reply.

This is my pandas dataframe 

![normal_df](https://cloud.githubusercontent.com/assets/1428796/22589176/0043d2ce-ea2f-11e6-9408-b08c9fbed6fe.PNG)

I converted the date column to float using the **[year] * 365 + [month] * 31 + [day]** and assigned it to **date_float** column.

![convert_df](https://cloud.githubusercontent.com/assets/1428796/22589235/4f4c6c64-ea2f-11e6-8ec9-e5513b236d42.PNG)

then used the following code 

```
from tpot import TPOTRegressor
from sklearn.model_selection import train_test_split
my_tpot = TPOTRegressor()
features = df['date_float']
outcome = df['metric']
X_train, X_test, y_train, y_test = train_test_split(pd.np.array(features), pd.np.array(outcome).ravel(),train_size=0.75, test_size=0.25)
tpot = TPOTRegressor(generations=5, population_size=20, verbosity=2)
tpot.fit(X_train, y_train)
```
where am getting the following error.

![error_format](https://cloud.githubusercontent.com/assets/1428796/22589346/db8e35c2-ea2f-11e6-9ee0-c06d30b82cb9.PNG)

Is there any problem in the way I passed the data ?

**Comment by rhiever** [comment]
You have to convert the DataFrame to a numpy array:

```
from tpot import TPOTRegressor
from sklearn.model_selection import train_test_split
my_tpot = TPOTRegressor()
features = df['date_float'].values
outcome = df['metric'].values
X_train, X_test, y_train, y_test = train_test_split(pd.np.array(features), pd.np.array(outcome).ravel(),train_size=0.75, test_size=0.25)
tpot = TPOTRegressor(generations=5, population_size=20, verbosity=2)
tpot.fit(X_train, y_train)
```

then it should work.

**Comment by AnnamalaiNagappan** [comment]
I copy pasted the code, still getting the same error.

![error_format](https://cloud.githubusercontent.com/assets/1428796/22596559/16652bca-ea52-11e6-8995-989d8a1e276c.PNG)

**Comment by weixuanfu2016** [comment]
Please try again after changing this line `X_train, X_test, y_train, y_test = train_test_split(pd.np.array(features).ravel(), pd.np.array(outcome).ravel(),train_size=0.75, test_size=0.25)` to `X_train, X_test, y_train, y_test = train_test_split(pd.np.array(features), pd.np.array(outcome).ravel(),train_size=0.75, test_size=0.25)`

Since X should be 2-D array but y is 1-D one.

**Comment by AnnamalaiNagappan** [comment]
Have changed the line, no use

![code_block](https://cloud.githubusercontent.com/assets/1428796/22615662/a9384faa-eabf-11e6-9904-ec39fe19e581.PNG)


I have added the data file and python notebook to this link 
[`https://github.com/AnnamalaiNagappan/tpot_share/blob/master/tpot_share.ipynb](url)
`

**Comment by rhiever** [comment]
Try taking out the `pd.np.array` calls in this line:

```
X_train, X_test, y_train, y_test = train_test_split(pd.np.array(features), pd.np.array(outcome).ra...
```

I suspect it's a data formatting issue still.

**Comment by weixuanfu2016** [comment]
@rhiever @AnnamalaiNagappan Based on the Jupyther notebook, I think the issue is caused by `features = df['date_float'].values`, which export a 1-D array so that X-train and X_test are a 1-D array. But X for `TPOT.fit` need to be a 2-D matrix. Please change `features = df['date_float'].values` to:
```
features = df['date_float'].values.reshape((df.shape[0],1))
```

**Comment by rhiever** [comment]
Good catch @weixuanfu2016.

**Comment by AnnamalaiNagappan** [comment]
Thank you very much. Now I am able to execute the code.

**Comment by rhiever** [comment]
Great! Let us know how it works.

**Comment by AnnamalaiNagappan** [comment]
I have updated the github repository [https://github.com/AnnamalaiNagappan/tpot_share/blob/master/tpot_share.ipynb](https://github.com/AnnamalaiNagappan/tpot_share/blob/master/tpot_share.ipynb)

Find the result i got 

![predict_output](https://cloud.githubusercontent.com/assets/1428796/22927389/fa651b00-f2d6-11e6-87f3-9822d2cc6ad5.PNG)

**Comment by rhiever** [comment]
Great! I would give it more time (i.e., larger population > 100 and more generations > 100) to see what it finds.

**Comment by rhiever** [comment]
We're closing all questions that haven't been updated in a while. Please feel free to re-open this issue if your issue persists.

**Comment by sonuthomas1996** [comment]
Hi, I'm trying out tpot for time series problem and i have two separate dataset, one for training and one for testing. While using normal train test split it gives you x_train, y_train, x_test, y_test. In this case im having separate dataset. Since i'm new to machine learning i'm confused then what are the parameters i need to pass for fit function , normally x_train and y_train is passed. In my case what should i pass? If someone could help me perform the timeseries it would be very helpful. My dataset contains 14 columns , one is the date and time and other is output variable and the related features.

**Comment by timkofu** [comment]
@rhiever Is there a way to integrate `tpot` with [`sktime`](https://www.sktime.org/en/latest/)?

**Comment by mach881040** [comment]
^
